src/insert_formatted.rs

```diff
@@ -0,0 +1,627 @@
+/// Otherwise, the whole `INSERT` will be aborted.
+///
+/// Rows are sent progressively to spread network load.
+#[must_use]
+pub struct InsertFormatted {
+    state: InsertState,
+    #[cfg(feature = "lz4")]
+    compression: Compression,
+    send_timeout: Option<Timeout>,
+    end_timeout: Option<Timeout>,
+    // Use boxed `Sleep` to reuse a timer entry, it improves performance.
+    // Also, `tokio::time::timeout()` significantly increases a future's size.
+    sleep: Pin<Box<Sleep>>,
+}
+
+struct Timeout {
+    duration: Duration,
+    is_set: bool,
+}
+
+enum InsertState {
+    NotStarted {
+        client: Box<Client>,
+        sql: String,
+    },
+    Active {
+        sender: ChunkSender,
+        handle: JoinHandle<Result<()>>,
+    },
+    Terminated {
+        handle: JoinHandle<Result<()>>,
+    },
+    Completed,
+}
+
+impl InsertState {
+    #[inline(always)]
+    fn is_not_started(&self) -> bool {
+        matches!(self, Self::NotStarted { .. })
+    }
+
+    fn sender(&mut self) -> Option<&mut ChunkSender> {
+        match self {
+            InsertState::Active { sender, .. } => Some(sender),
+            _ => None,
+        }
+    }
+
+    fn handle(&mut self) -> Option<&mut JoinHandle<Result<()>>> {
+        match self {
+            InsertState::Active { handle, .. } | InsertState::Terminated { handle } => Some(handle),
+            _ => None,
+        }
+    }
+
+    fn client_with_sql(&self) -> Option<(&Client, &str)> {
+        match self {
+            InsertState::NotStarted { client, sql } => Some((client, sql)),
+            _ => None,
+        }
+    }
+
+    #[inline]
+    fn expect_client_mut(&mut self) -> &mut Client {
+        let Self::NotStarted { client, .. } = self else {
+            panic!("cannot modify client options while an insert is in-progress")
+        };
+
+        client
+    }
+
+    fn terminated(&mut self) {
+        match mem::replace(self, InsertState::Completed) {
+            InsertState::NotStarted { .. } | InsertState::Completed => (),
+            InsertState::Active { handle, .. } => {
+                *self = InsertState::Terminated { handle };
+            }
+            InsertState::Terminated { handle } => {
+                *self = InsertState::Terminated { handle };
+            }
+        }
+    }
+}
+
+impl InsertFormatted {
+    pub(crate) fn new(client: &Client, sql: String) -> Self {
+        Self {
+            state: InsertState::NotStarted {
+                client: Box::new(client.clone()),
+                sql,
+            },
+            #[cfg(feature = "lz4")]
+            compression: client.compression,
+            send_timeout: None,
+            end_timeout: None,
+            sleep: Box::pin(tokio::time::sleep(Duration::new(0, 0))),
+        }
+    }
+
+    /// Sets timeouts for different operations.
+    ///
+    /// `send_timeout` restricts time on sending a data chunk to a socket.
```

Comment:
docs CI probably will complain about this without square brackets

Reply:
This was copied straight from `Insert`